thread_timeout/__init__.py

```diff
@@ -101,84 +103,145 @@ class NotKillExecTimeoutException(ExecTimeoutException):
-            return result[0]
-        return decorator
-    return generate_decorator
-
-if __name__ == "__main__":
-    print ("Running tests")
-
+    @wrapt.decorator
+    def wrapper(wrapped, instance, args, kwargs):
+        queue = Queue()
+
+        def inner_worker():
+            result = wrapped(*args, **kwargs)
+            queue.put(result)
+        thread = threading.Thread(target=inner_worker)
+        thread.daemon = True
+        thread.start()
+        thread.join(delay)
+        if thread.isAlive():
+            if not kill:
+                raise NotKillExecTimeoutException(
+                    "Timeout and no kill attempt")
+            _kill_thread(thread)
+            time.sleep(kill_wait)
+            ## FIXME isAlive is giving fals positive results
+            # if thread.isAlive():
+            #     raise FailedKillExecTimeoutException(
+            #         "Timeout, thread refuses to die in %s seconds" %
+            #         kill_wait)
+            # else:
+            raise KilledExecTimeoutException(
+                    "Timeout and thread was killed")
+        return queue.get()
+    return wrapper
+
+
+def test1():
+    ''' timeout is not stopping quick function
+    '''
     @thread_timeout(2)
-    def test1(delay):
-        print ("  .. sleeing for %s" % delay)
+    def func(delay):
+        print("  .. sleeing for %s" % delay)
         time.sleep(delay)
-        print ("  .. done sleep for %s" % delay)
+        print("  .. done sleep for %s" % delay)
     try:
-        res = test1(1)
-        print ("Test1 OK")
+        res = func(1)
+        print("Test1 OK")
     except ExecTimeoutException:
-        print ("Test1 failed, timeout too soon")
+        print("Test1 failed, timeout too soon")
 
+
+def test2():
+    ''' timeout is stopping long function
+    '''
     @thread_timeout(1)
-    def test2(delay):
-        print ("  .. sleeing for %s" % delay)
+    def func(delay):
+        print("  .. sleeing for %s" % delay)
         time.sleep(delay)
         print("  .. done sleep for %s" % delay)
 
     try:
-        test2(3)
+        func(3)
         raise Exception("Test2 failed: timeout does not work")
     except ExecTimeoutException as e:
-        print ("  .. got excepted execption %s" % repr(e))
+        print("  .. got excepted execption %s" % repr(e))
         print("Test2 OK")
 
+
+def test3():
+    ''' function returns result
+    '''
     @thread_timeout(1)
-    def test3(x):
+    def func(x):
         return x
 
-    if test3('OK') != 'OK':
-        print ("Test3 failed, %s", repr(test3('OK')))
-    else:
-        print ("Test3 OK")
+    assert func('OK') == 'OK'
+
 
+def test4():
+    ''' FailedKillExecTimeoutException
+    FIXME! This is a wired test.  thread_timeout should actually stop this function 
+    '''
     @thread_timeout(1)
-    def test4(x):
+    def looong(x):
         for a in range(0, x):
             time.sleep(2)
     try:
-        test4(2)
+        looong(20000)
```

Comment:
btw look here https://gist.github.com/iakovgan/058a597f6beab87c0068
i used time measurements for testing of a similar module. you might find it useful